### PR TITLE
Stop testing EOL versions of node

### LINKED ISSUE #1038: Drop support for node versions that are EOL
I tried updating some devDependencies in #1035 and ran into test failures in node v6 when newer libraries tried to use the spread operator - `...`

Node v6 has been [EOL](https://github.com/nodejs/Release) for about 5 months now. It would be great if we can drop support for v6 so that we can start using modern javascript and libraries written in modern javascript.

I see tests are running on 6, 7, 8, 9, 10, 11, and 12. Only 8, 10 and 12 are not EOL now.

My proposal is to only run tests on 8, 10 and 12 as they are the only supported versions of node at the moment.

### PR DESCRIPTION
**Pull request by bjornstar**
Resolves #1038 

In order to use modern javascript syntax and libraries using that syntax we need to stop supporting node v6. At the same time, we can also stop testing other versions of node that are officially EOL.

This PR removes testing against node v6, 7, 9, and 11 while moving the initial node version from v8 to v10.

I also added `node-gyp` as a devDependency since it is required to build.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -50,6 +51,9 @@
       "name": "Google Inc."
     }
   ],
+  "engines": {
+    "node": "^8.13.0 || >=10.10.0"
```
Comment (murgatroid99) [inline package.json]: I don't really see a point to this, because this package isn't published and this range isn't binding anyway.

File: run-tests.bat [hunk 2]
```diff
@@ -24,12 +24,12 @@ powershell -c "(Get-WmiObject -Class Win32_ComputerSystem).SystemType"
 powershell -c "& { iwr https://raw.githubusercontent.com/grumpycoders/nvm-ps/master/nvm.ps1 | iex }"
 
 SET PATH=%APPDATA%\nvm-ps;%APPDATA%\nvm-ps\nodejs;%PATH%
-SET JOBS=8
+SET JOBS=4
```
Comment (murgatroid99) [inline run-tests.bat]: What is the purpose of this change?

File: package.json (hunk 1 above)
Comment (bjornstar) [reply package.json]: It's advisory the same way it is in `grpc-js`, I've removed it.

**Comment by murgatroid99** [comment]
A dev dependency on node-gyp shouldn't be necessary. It comes with npm.

**Comment by bjornstar** [comment]
> A dev dependency on node-gyp shouldn't be necessary. It comes with npm.

It doesn't It's being installed by the test scripts so I thought to include it. I'll move it into another PR since it's unrelated to this change.

**Comment by murgatroid99** [comment]
node-gyp does come with npm. We install it in the test script to work around a bug in old versions of npm that are included by default in some versions of node that we test, including node 8. But it is not always necessary, and as new node versions are released it should be necessary in fewer cases.

**Comment by murgatroid99** [comment]
That last commit is not acceptable. We still want to publish binaries for older versions of Node.

**Comment by bjornstar** [comment]
> That last commit is not acceptable. We still want to publish binaries for older versions of Node.

Understood, I've backed out all the changes except which node versions to test on.
